test/jdk/java/lang/annotation/LoaderLeakTest.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2004, 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 5040740
+ * @summary annotations cause memory leak
+ * @library /test/lib
+ * @build jdk.test.lib.process.*
+ * @run testng LoaderLeakTest
+ */
+
+import jdk.test.lib.Utils;
+import jdk.test.lib.process.ProcessTools;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+import java.io.FileInputStream;
+import java.lang.annotation.Retention;
+import java.lang.ref.Reference;
+import java.lang.ref.WeakReference;
+import java.nio.file.*;
+import java.util.*;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+public class LoaderLeakTest {
+
+    @Test
+    public void testWithoutReadingAnnotations() throws Throwable {
+        runJavaProcessExpectSuccessExitCode("Main");
+    }
+
+    @Test
+    public void testWithReadingAnnotations() throws Throwable {
+        runJavaProcessExpectSuccessExitCode("Main",  "foo");
+    }
+
+    private void runJavaProcessExpectSuccessExitCode(String ... command) throws Throwable {
+        var processBuilder = ProcessTools.createJavaProcessBuilder(command)
+                                                      .directory(Paths.get(Utils.TEST_CLASSES).toFile());
+        ProcessTools.executeCommand(processBuilder).shouldHaveExitValue(0);
+    }
+
+}
+
+class Main {
+
+    public static void main(String[] args) throws Exception {
+        for (int i = 0; i < 100; i++) {
+            try {
+                doTest(args.length != 0);
+            }
+            catch (NullPointerException npe) {
+                throw new AssertionError("c.get() should never return null", npe);
+            }
```

Comment:
I don't think it's the right way to handle that, you don't know if this NPE is from `c.get`, so the exception messages might be misleading. I'd just remove try/catch, if NPE occurs jtreg will report the test as failed w/ NPE as a reason, so whoever analyzes it will be able to understand.

alternatively, you can save c.get into a local variable which you nulls later one, e.g.
 ```
static void doTest(boolean readAnn) throws Exception {
...
var tmp = c.get();
if (t == null) throw new AssertionError("c.get is null");
if (t.getClassLoader() != loader) throw new AssertionError("wrong classloader");
t = null;
```

Reply:
@iignatev I just reemove the try catch I think the comments are informative and will help with analysis with potential NPE